### PR TITLE
Issue1926 actuator travel

### LINKED ISSUE #1926: Refactor actuator travel time
The current implementation of the actuator model, using the 2nd order filter, causes always similar time to make a set point change, regardless of the size of the set point change. A more realistic implementation would use a slew rate limiter.

This issue is to test the performance of using a slew rate limiter and if the performance is good, change the actuator model that is used for pumps, valves and dampers.

The implementation is also done at https://github.com/lbl-srg/modelica-buildings/issues/3965 to allow for benchmarking across a larger set of test cases.

The development branch is `issue1926_actuatorTravel`

### PR DESCRIPTION
**Pull request by mwetter**
This closes #1926

### CODE REVIEW

**Comment by mwetter** [comment]
This is ready for review and merge.
@jelgerjansen : Can you please review.
@nytschgeusen @FWuellhorst : Can you please give your OK within 2 weeks as this is a non-backward compatible change. The conversion script has been updated.

File: IBPSA/Fluid/Movers/SpeedControlled_y.mo [hunk 1]
```diff
@@ -6,19 +6,18 @@ model SpeedControlled_y
...
       y(final unit="1")),
-    eff(
-      per(final pressure = per.pressure,
-          final etaHydMet = per.etaHydMet,
-          final etaMotMet = per.etaMotMet),
-      r_N(start=y_start)));
+    eff(per(
+        final pressure=per.pressure,
+        final etaHydMet=per.etaHydMet,
+        final etaMotMet=per.etaMotMet), r_N(start=y_start)));
```
Comment (jelgerjansen) [inline IBPSA/Fluid/Movers/SpeedControlled_y.mo:17]: Move `r_N(...` to new line to improve readability.

File: IBPSA/Fluid/Movers/Validation/ComparePowerHydraulic.mo [hunk 2]
```diff
@@ -12,7 +12,7 @@ model ComparePowerHydraulic
       per=per,
       energyDynamics=Modelica.Fluid.Types.Dynamics.SteadyState,
       addPowerToMedium=false,
-      use_inputFilter=false),
+      use_riseTime   =false),
```
Comment (jelgerjansen) [inline IBPSA/Fluid/Movers/Validation/ComparePowerHydraulic.mo:15]: Remove abundant whitespace after `use_riseTime`.

File: IBPSA/Fluid/Movers/Validation/ComparePowerHydraulic.mo [hunk 3]
```diff
@@ -23,7 +23,7 @@ model ComparePowerHydraulic
         etaMotMet=per.etaMotMet),
       energyDynamics=Modelica.Fluid.Types.Dynamics.SteadyState,
       addPowerToMedium=false,
-      use_inputFilter=false),
+      use_riseTime   =false),
```
Comment (jelgerjansen) [inline IBPSA/Fluid/Movers/Validation/ComparePowerHydraulic.mo:26]: Remove abundant whitespace after `use_riseTime`.

File: IBPSA/Fluid/Movers/Validation/ComparePowerHydraulic.mo [hunk 4]
```diff
@@ -34,7 +34,7 @@ model ComparePowerHydraulic
         etaMotMet=per.etaMotMet),
       energyDynamics=Modelica.Fluid.Types.Dynamics.SteadyState,
       addPowerToMedium=false,
-      use_inputFilter=false));
+      use_riseTime   =false));
```
Comment (jelgerjansen) [inline IBPSA/Fluid/Movers/Validation/ComparePowerHydraulic.mo:37]: Remove abundant whitespace after `use_riseTime`.

File: IBPSA/Fluid/Movers/Validation/ComparePowerInput.mo [hunk 5]
```diff
@@ -11,23 +11,23 @@ model ComparePowerInput
...
     redeclare IBPSA.Fluid.Movers.FlowControlled_m_flow mov3(
       redeclare final package Medium = Medium,
       energyDynamics=Modelica.Fluid.Types.Dynamics.SteadyState,
       per=per,
       addPowerToMedium=false,
-      use_inputFilter=false,
       m_flow_nominal=m_flow_nominal,
-      dp_nominal=dp_nominal),
+      dp_nominal=dp_nominal,
+      use_riseTime   =false),
```
Comment (jelgerjansen) [inline IBPSA/Fluid/Movers/Validation/ComparePowerInput.mo:30]: Remove abundant whitespace after `use_riseTime`.

File: IBPSA/Fluid/Movers/Validation/ControlledFlowMachine.mo [hunk 6]
```diff
@@ -5,12 +5,12 @@ model ControlledFlowMachine "Fans with different control signals as input"
...
       addPowerToMedium=false,
-      use_inputFilter=false,
+      use_riseTime=false,
       energyDynamics=Modelica.Fluid.Types.Dynamics.FixedInitial),
     fan3(
       addPowerToMedium=false,
-      use_inputFilter=false,
-      energyDynamics=Modelica.Fluid.Types.Dynamics.FixedInitial));
+      energyDynamics=Modelica.Fluid.Types.Dynamics.FixedInitial,
+      use_riseTime   =false));
```
Comment (jelgerjansen) [inline IBPSA/Fluid/Movers/Validation/ControlledFlowMachine.mo:13]: Remove abundant whitespace after `use_riseTime`.

File: IBPSA/Fluid/Movers/Validation/FlowControlled_dp.mo [hunk 7]
```diff
@@ -7,14 +7,14 @@ model FlowControlled_dp "Fan with zero mass flow rate and head as input"
...
+      energyDynamics=Modelica.Fluid.Types.Dynamics.SteadyState,
+      use_riseTime   =false),
     redeclare IBPSA.Fluid.Movers.FlowControlled_dp floMacDyn(
       nominalValuesDefineDefaultPressureCurve=true,
       redeclare package Medium = Medium,
       m_flow_nominal=m_flow_nominal,
-      use_inputFilter=false,
-      energyDynamics=Modelica.Fluid.Types.Dynamics.FixedInitial));
+      energyDynamics=Modelica.Fluid.Types.Dynamics.FixedInitial,
+      use_riseTime   =false));
```
Comment (jelgerjansen) [inline IBPSA/Fluid/Movers/Validation/FlowControlled_dp.mo:17]: Remove abundant whitespace after `use_riseTime`.

File: IBPSA/Fluid/Movers/Validation/FlowControlled_m_flow.mo [hunk 8]
```diff
@@ -8,14 +8,14 @@ model FlowControlled_m_flow
...
+      energyDynamics=Modelica.Fluid.Types.Dynamics.SteadyState,
+      use_riseTime   =false),
     redeclare IBPSA.Fluid.Movers.FlowControlled_m_flow floMacDyn(
       nominalValuesDefineDefaultPressureCurve=true,
       redeclare package Medium = Medium,
       m_flow_nominal=m_flow_nominal,
-      use_inputFilter=false,
-      energyDynamics=Modelica.Fluid.Types.Dynamics.FixedInitial));
+      energyDynamics=Modelica.Fluid.Types.Dynamics.FixedInitial,
+      use_riseTime   =false));
```
Comment (jelgerjansen) [inline IBPSA/Fluid/Movers/Validation/FlowControlled_m_flow.mo:18]: Remove abundant whitespace after `use_riseTime`.

File: IBPSA/Fluid/Movers/Validation/Pump_stratos.mo [hunk 9]
```diff
@@ -8,12 +8,12 @@ model Pump_stratos "Stratos pumps with speed as input"
...
-      use_inputFilter=false,
-      per=per),
+      per=per,
+      use_riseTime   =false),
     redeclare IBPSA.Fluid.Movers.SpeedControlled_y floMacDyn(
       redeclare package Medium = Medium,
-      use_inputFilter=false,
-      per=per));
+      per=per,
+      use_riseTime   =false));
```
Comment (jelgerjansen) [inline IBPSA/Fluid/Movers/Validation/Pump_stratos.mo:16]: Remove abundant whitespace after `use_riseTime`.

File: IBPSA/Fluid/Movers/Validation/SpeedControlled_y.mo [hunk 10]
```diff
@@ -8,14 +8,14 @@ model SpeedControlled_y
...
+      energyDynamics=Modelica.Fluid.Types.Dynamics.SteadyState,
+      use_riseTime   =false),
     redeclare IBPSA.Fluid.Movers.SpeedControlled_y floMacDyn(
       redeclare package Medium = Medium,
       per(pressure(V_flow={0,m_flow_nominal,2*m_flow_nominal}/1.2,
                     dp={2*dp_nominal,dp_nominal,0})),
-      use_inputFilter=false,
-      energyDynamics=Modelica.Fluid.Types.Dynamics.FixedInitial));
+      energyDynamics=Modelica.Fluid.Types.Dynamics.FixedInitial,
+      use_riseTime   =false));
```
Comment (jelgerjansen) [inline IBPSA/Fluid/Movers/Validation/SpeedControlled_y.mo:18]: Remove abundant whitespace after `use_riseTime`.

**Review by jelgerjansen — CHANGES_REQUESTED** [review]
@mwetter thank you for your nice work! Most of my feedback are minor comments. Two issues that might require some more time:

- I noticed that the renaming (conversion script) also changed variable names in the revision history, which is unintended. I'm not sure if this can be avoided somehow?
- The value of `Td` (in the `SlewRateLimiter`) is defined as `10/strokeTime`, but I would have expected `strokeTime` in the numerator (both in terms of units and physical meaning). Could you explain how you choose the expression for `Td`?

File: IBPSA/Fluid/Actuators/BaseClasses/ActuatorSignal.mo [hunk 11]
```diff
@@ -33,37 +34,36 @@ model ActuatorSignal
...
-    final normalized=true,
-    final initType=init,
-    final y_start=y_start) if use_inputFilter
-    "Second order filter to approximate actuator opening time, and to improve numerics"
-    annotation (Placement(transformation(extent={{6,81},{20,95}})));
-
+  Modelica.Blocks.Nonlinear.SlewRateLimiter actPos(
+    Rising=1/strokeTime,
+    Falling=-1/strokeTime,
+    Td=10/strokeTime,
```
[the remark this answers is not in the document]
Comment (mwetter) [reply IBPSA/Fluid/Actuators/BaseClasses/ActuatorSignal.mo]: This is now corrected on this development branch through https://github.com/ibpsa/modelica-ibpsa/pull/1933/commits/1ac4b554d2ffed41651b18c40a7a5fabb08b4ff2

File: IBPSA/Electrical/Transmission/BaseClasses/BaseCable.mo [hunk 12]
```diff
@@ -17,10 +17,16 @@ record BaseCable "Record that contains the properties of a generic cable"
...
+    Modelica.Units.SI.AngularVelocity omega=2*Modelica.Constants.pi*f;
     annotation (Documentation(revisions="<html>
 <ul>
 <li>
+August 30, 2024, by Michael Wetter:<br/>
+Removed wrong parameter keyword on protected variable, which is needed for Dymola 2025x.<br/>
+This is for
+<a href=\"https://github.com/ibpsa/modelica-ibpsa/issues/1930\">IBPSA, #1930</a>.
+</li>
+<li>
```
[the remark this answers is not in the document]
Comment (mwetter) [reply IBPSA/Electrical/Transmission/BaseClasses/BaseCable.mo:29]: Corrected by adding both revision notes.

File: IBPSA/Fluid/Actuators/BaseClasses/ActuatorSignal.mo [hunk 13]
```diff
@@ -119,6 +120,11 @@ for a description of the filter.
 </html>", revisions="<html>
 <ul>
 <li>
+August 26, 2024, by Michael Wetter:<br/>
+Implemented linear actuator travel dynamics.<br/>
+This is for <a href=\"https://github.com/lbl-srg/modelica-buildings/issues/3965\">Buildings, #3965</a>.
```
[the remark this answers is not in the document]
Comment (mwetter) [reply IBPSA/Fluid/Actuators/BaseClasses/ActuatorSignal.mo:125]: Done, both are mentioned now.

File: IBPSA/Fluid/Actuators/BaseClasses/PartialThreeWayValve.mo [hunk 14]
```diff
@@ -1,36 +1,33 @@
...
-          final CvData=IBPSA.Fluid.Types.CvTypes.OpPoint,
-          final m_flow_nominal=m_flow_nominal,
-          final dpValve_nominal=dpValve_nominal/fraK^2,
-          final dpFixed_nominal=dpFixed_nominal[2],
-          final use_inputFilter=false,
-          final riseTime=riseTime));
+    m_flow_small=m_flow_nominal*1e-4,
+    final mDyn_flow_nominal=m_flow_nominal,
+    redeclare replaceable IBPSA.Fluid.Actuators.BaseClasses.PartialTwoWayValve
+      res1 constrainedby IBPSA.Fluid.Actuators.BaseClasses.PartialTwoWayValve(
```
[the remark this answers is not in the document]
Comment (mwetter) [reply IBPSA/Fluid/Actuators/BaseClasses/PartialThreeWayValve.mo:7]: Done.

File: IBPSA/Fluid/Movers/Validation/ControlledFlowMachine.mo (hunk 6 above)
Comment (mwetter) [reply IBPSA/Fluid/Movers/Validation/ControlledFlowMachine.mo:13]: Done.

**Comment by mwetter** [comment]
@FWuellhorst : Thanks for the review. All changes have been addressed.

File: IBPSA/Fluid/Actuators/BaseClasses/ActuatorSignal.mo (hunk 13 above)
[the remark this answers is not in the document]
Comment (jelgerjansen) [reply IBPSA/Fluid/Actuators/BaseClasses/ActuatorSignal.mo]: I noticed an abundant whitespace before the IBPSA hyperlink, which I fixed through commit 27c8c85

**Comment by jelgerjansen** [comment]
@mwetter I think you indeed mixed up the names of @FWuellhorst and myself ;)

Thanks for fixing the time constant of the SlewRateLimiter block. Did this lead to significant changes in simulation time and/or results?

The updated revision notes contained an abundant white space, which I fixed myself through commit 27c8c85.

Furthermore, I noticed that some of my remarks were not yet addressed. Did you read over these comments or do you prefer the implementation as is?

File: IBPSA/Fluid/Actuators/BaseClasses/ActuatorSignal.mo [hunk 15]
```diff
@@ -33,37 +34,36 @@ model ActuatorSignal
...
     annotation(Evaluate=true);
   Modelica.Blocks.Interfaces.RealOutput y_internal(unit="1")
     "Output connector for internal use (= y_actual if not casePreInd)";
-  Modelica.Blocks.Interfaces.RealOutput y_filtered if use_inputFilter
+  Modelica.Blocks.Interfaces.RealOutput y_filtered if use_strokeTime
     "Filtered valve position in the range 0..1"
     annotation (Placement(transformation(extent={{40,78},{60,98}}),
         iconTransformation(extent={{60,50},{80,70}})));
 
-  IBPSA.Fluid.BaseClasses.ActuatorFilter filter(
```
[the remark this answers is not in the document]
Comment (mwetter) [reply IBPSA/Fluid/Actuators/BaseClasses/ActuatorSignal.mo:50]: Let's move it through a separate PR. I opened https://github.com/ibpsa/modelica-ibpsa

**Comment by mwetter** [comment]
@jelgerjansen : Sorry for missing some comments, I thought I addressed them through the text replace.

The simulation time stays about the same, some models become faster and others slower, see the link in the issue. In `Buildings` we have some models that don't work and we are looking into this, the reason being pump working against a closed valve and some chattering in control logic. I think it is not fundamental to this change but rather a change in numerics that now needs retuning or reformulation of some models that we got to work for the old filter model. I would expect a similar change being needed if models would have been developed for the new SlewRateLimiter and we were to change to the 2nd order filter.

**Comment by jelgerjansen** [comment]
@mwetter thanks for addressing the other comments as well. I think this is now ready to be merged.

**Review by jelgerjansen — APPROVED** [review]

**Pull request merged**